Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +133,757 @@ final class ConnectionPool {
         }
+    }
 
-        var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+    /// Removes idle timeout handler and channel inactivity listener
+    func cancelIdleTimeout() -> EventLoopFuture<Void> {
+        return self.removeHandler(IdleStateHandler.self).flatMap { _ in
+            self.removeHandler(IdlePoolConnectionHandler.self)
         }
+    }
+}
 
-        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
-            switch action {
-            case .leaseConnection(let connection):
-                return connection.removeIdleConnectionHandlersForLease().flatMapError { _ in
-                    connection.closeFuture.flatMap { // We ensure close actions are run first
-                        let defaultEventLoop = self.stateLock.withLock {
-                            self.state.defaultEventLoop
-                        }
-                        return self.makeConnection(on: preference.bestEventLoop ?? defaultEventLoop)
-                    }
-                }
-            case .makeConnection(let eventLoop):
-                return self.makeConnection(on: eventLoop)
-            case .leaseFutureConnection(let futureConnection):
-                return futureConnection
-            }
+extension Connection: Hashable {
+    static func == (lhs: Connection, rhs: Connection) -> Bool {
+        return ObjectIdentifier(lhs) == ObjectIdentifier(rhs)
+    }
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(ObjectIdentifier(self))
+    }
+}
+
+/// A connection provider of `HTTP/1.1` connections with a given `Key` (host, scheme, port)
+///
+/// On top of enabling connection reuse this provider it also facilitates the creation
+/// of concurrent requests as it has built-in politeness regarding the maximum number
+/// of concurrent requests to the server.
+class HTTP1ConnectionProvider: CustomStringConvertible {
+    enum Action {
+        case lease(Connection, Waiter)
+        case create(Waiter)
+        case replace(Connection, Waiter)
+        case closeProvider
+        case park(Connection)
+        case none
+        case fail(Waiter, Error)
+        case cancel(Connection, Bool)
+        indirect case closeAnd(Connection, Action)
+        indirect case parkAnd(Connection, Action)
+    }
+
+    struct ConnectionsState {
+        enum State {
+            case active
+            case closed
         }
 
-        func release(connection: Connection) {
-            self.activityPrecondition(expected: [.opened, .closing])
-            let action = self.parentPool.connectionProvidersLock.withLock {
-                self.stateLock.withLock { self.state.releaseAction(for: connection) }
+        #if DEBUG
```

Comment:
this won't work if you run tests in release mode, I think we should just remove it.

Reply:
I don't think you can run our tests in release mode at all, at least most of them, due to `@testable` import, no? but, yeah, I'll remove it, I just wanted to make sure those methods cannot be used in the "real" code... maybe it was a bit of an overreaction 😄

Reply:
Yes, you can :) `swift build -c release -Xswiftc -enable-testing`. The NIO release tools run them for every project in the following confugurations:

- debug
- ASan in _release_ mode (ASan works MUCH better in Swift when run in release mode)
- TSan in debug mode
- in docker in debug mode

Reply:
huh, did not know that, don't usually run tests from command line 😆

Reply:
done